discovery_engine_core/providers/src/expression.rs

```diff
@@ -0,0 +1,376 @@
+        let a = Expr::Value("a".to_string());
+
+        assert_eq!(a, empty.clone().or(a.clone()));
+
+        let one = Expr::Or(btreeset![a.clone()]);
+        assert_eq!(a, empty.or(one));
+    }
+
+    #[test]
+    fn test_or_and() {
+        let a = Expr::Value("a".to_string());
+        let b = Expr::Value("b".to_string());
+        let c = Expr::Value("c".to_string());
+        let expected = Expr::Or(btreeset![
+            a.clone(),
+            Expr::And(btreeset![b.clone(), c.clone()])
+        ]);
+
+        // a \/ (b /\ c)
+        assert_eq!(expected, a.clone().or(b.clone().and(c.clone())));
+
+        // (b /\ c) \/ a
+        assert_eq!(expected, b.and(c).or(a));
+    }
+
+    #[test]
+    fn test_build_value_empty() {
+        let empty = "".to_string();
+
+        assert_eq!(empty, Expr::Value(empty.clone()).build());
+    }
+
+    #[test]
+    fn test_build_value() {
+        let a = "a".to_string();
+
+        assert_eq!(a, Expr::Value(a.clone()).build());
+    }
+
+    #[test]
+    fn test_build_and_empty() {
+        let v = Expr::Value("".to_string());
+
+        let empty_and = v.clone().and(v.clone());
+        assert_eq!("", empty_and.build());
+
+        let empty_and_empty_and = v.clone().or(v.clone().or(v.clone()));
+        assert_eq!("", empty_and_empty_and.build());
+
+        let empty_and_empty_or = v.clone().and(v.clone().or(v));
+        assert_eq!("", empty_and_empty_or.build());
+    }
+
+    #[test]
+    fn test_build_and_values() {
+        let a = Expr::Value("a".to_string());
+        let b = Expr::Value("b".to_string());
+        let c = Expr::Value("c".to_string());
+
+        assert_eq!(
+            "a AND b AND c",
+            a.clone().and(b.clone()).and(c.clone()).build()
+        );
+        assert_eq!("a AND b AND c", c.and(b.and(a)).build());
+    }
+
+    #[test]
+    fn test_build_or_empty() {
+        let v = Expr::Value("".to_string());
+
+        let empty_or = v.clone().or(v.clone());
+        assert_eq!("", empty_or.build());
+
+        let empty_or_empty_or = v.clone().or(v.clone().or(v.clone()));
+        assert_eq!("", empty_or_empty_or.build());
+
+        let empty_or_empty_and = v.clone().or(v.clone().and(v));
+        assert_eq!("", empty_or_empty_and.build());
+    }
+
+    #[test]
+    fn test_build_or_values() {
+        let a = Expr::Value("a".to_string());
+        let b = Expr::Value("b".to_string());
+        let c = Expr::Value("c".to_string());
+
+        assert_eq!("a OR b OR c", a.clone().or(b.clone()).or(c.clone()).build());
+        assert_eq!("a OR b OR c", c.or(b.or(a)).build());
+    }
+
+    #[test]
+    fn test_build_and_or() {
+        let a = Expr::Value("a".to_string());
+        let b = Expr::Value("b".to_string());
+        let c = Expr::Value("c".to_string());
+        let d = Expr::Value("d".to_string());
+
+        // (a /\ b) \/ c
+        assert_eq!(
+            "(a AND b) OR c",
+            a.clone().and(b.clone()).or(c.clone()).build()
```

Comment:
Are these expressions that we would directly pass to WebZ? Or who will be the consumer for these strings?

Reply:
The endpoint what the string in this format.